_drafts/assert.md

```diff
@@ -0,0 +1,605 @@
+---
+title: "Using Asserts in Embedded Systems"
+description:
+  "Asserts are a quick and proven way to prevent bugs in embedded systems. An
+  optimal assert implementation both maximizes the context around the assert
+  while minimizing code size and RAM usage."
+author: tyler
+image: /img/debugging-asserts/assert.png
+---
+
+The use of asserts is one of the best ways to find bugs, unintended behavior,
+programmatic errors, and to learn when systems are no longer 100% functional and
+need to be reset. If instrumented correctly, an assert can give a developer
+context about when and where in the code an issue took place. Despite the
+numerous benefits, the practice of using asserts in firmware is not common or
+agreed upon.
+
+<!-- excerpt start -->
+
+By using asserts proactively in embedded systems on debug and production builds,
+developers can both prevent bugs before shipping and quickly surface and fix
+them after shipping. Proper assert handling is also the safest way to handle
+issues and undefined behavior that occur in production. In this post, we'll go
+over best practices with asserts, when to use asserts, and then come up with a
+production ready custom assert implementation for an ARM Cortex-M device, all
+while keeping the code size usage to a minimum.
+
+<!-- excerpt end -->
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest posts
+straight to your mailbox_
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Asserts in Embedded Systems
+
+The use of asserts is controversial in the embedded space, **especially** in
```

Comment:
I would remove the `Asserts in Embedded Systems` section entirely. It reads like a second introduction and doesn't add meaningful information. 

Instead, add a note (one of those grey boxes) about "why you shouldn't compile asserts out in production builds".

It could say something like this:

> **Why you shouldn't compile asserts out in production builds**
> Many of the popular embedded platforms have options to compile out error handling and assertions. In our opinion this is a mistake. In the event of inconsistent behavior or memory corruption, crashing the device is often the safest thing you could do. Embedded systems can often reboot quickly, getting the system back to a known good state. Over time, it will also be noticed and provide valuable feedback back to your engineering teams. The alternative is worse: the system could behave in unpredictable way and perform poorly, lose customer information, ...etc.

Reply:
Awesome! Love this. Done